docs/tutorials/sparse/csr.md

```diff
@@ -0,0 +1,258 @@
+```
+
+* If the storage types of source array and destination array doesn't match,
+the storage type of destination array won't change when copying with `copyto` or
+the slice operator `[]`.
+
+```python
+e = mx.nd.sparse.zeros('csr', (2,2))
+f = mx.nd.sparse.zeros('csr', (2,2))
+g = mx.nd.ones(e.shape)
+e[:] = g
+g.copyto(f)
+{'e.stype':e.stype, 'f.stype':f.stype}
+```
+
+## Indexing and Slicing
+
+We can slice a CSRNDArray on axis 0 with operator `[]`, which copies the slices and returns a new CSRNDArray.
+
+```python
+a = mx.nd.array(np.arange(6).reshape(3,2)).tostype('csr')
+b = a[1:2]
+c = a[:].asnumpy()
+{'b':b, 'c':c}
+```
+
+## Sparse Operators and Storage Type Inference
+
+* Operators that have specialized implementation for sparse arrays can be accessed in ``mx.nd.sparse``.
+You can read the [mxnet.ndarray.sparse API documentation](mxnet.io/api/python/ndarray.html) to find
+what sparse operators are available.
+
+```python
+shape = (3, 4)
+data = [7, 8, 9]
+indptr = [0, 2, 2, 3]
+indices = [0, 2, 1]
+a = mx.nd.sparse.csr_matrix(data, indptr, indices, shape) # a csr matrix as lhs
+rhs = mx.nd.ones((4, 1))      # a dense vector as rhs
+out = mx.nd.sparse.dot(a, rhs)  # invoke sparse dot operator specialized for dot(csr, dense)
+{'out':out}
+```
+
+* For any sparse operator, the storage type of output array is inferred based on inputs. You can either read
+the documentation or inspect the `stype` attribute of output array to know what storage type is inferred:
+
+```python
+b = a * 2  # b will be a CSRNDArray since zero multiplied by 2 is still zero
+c = a + 1  # c will be a dense NDArray
+{'b.stype':b.stype, 'c.stype':c.stype}
+```
+
+* For operators that don't specialize in sparse arrays, we can still use them with sparse inputs with some performance penalty.
+What happens is that MXNet will generate temporary dense inputs from sparse inputs so that the dense operators can be used.
+Warning messages will be printed when such storage fallback event happens.
+
+```python
+d = mx.nd.log(a) # warnings will be printed
+{'a.stype':a.stype, 'd':d} # stype of a is not changed
+```
+
+## Loading LibSVM Data
+
+Sparse data stored in libsvm file format can be loaded with [mx.io.LibSVMIter](https://mxnet.incubator.apache.org/versions/master/api/python/io.html#mxnet.io.LibSVMIter).
+Note that the indices are expected to be zero-based instead of one-based.
+
+## Advanced Topics
+
+### GPU Support
+
+By default, CSRNDArray operators are executed on CPU. In MXNet, GPU support for CSRNDArray is experimental
+with few sparse operators such as cast_storage and dot.
+
+To create a CSRNDArray on gpu, we need to explicitly specify the context:
+
+**Note** In order to execute the following section on a cpu set gpu_device to mx.cpu().
+```python
+gpu_device=mx.gpu() # Change this to mx.cpu() in absence of GPUs.
+
+a = mx.nd.sparse.zeros('csr', (100, 100), ctx=gpu_device)
+a
```

Comment:
If I run this code on macOS with no GPU, the python session seg-faults. I know that the context is set incorrectly to GPU when GPU is not present, but should the python session seg-fault? Shouldn't the python session give an error/exception that can be caught by the user and handled appropriately?

Reply:
Did it segfault and exit python? For me there was error msg "GPU support is disabled..": 

```
>>> mx.nd.sparse.zeros('csr', (100, 100), ctx=mx.gpu())
[20:01:08] src/c_api/c_api_ndarray.cc:148: GPU support is disabled. Compile MXNet with USE_CUDA=1 to enable GPU support.
[20:01:08] /Users/haibilin/mxnet/dmlc-core/include/dmlc/logging.h:308: [20:01:08] src/c_api/c_api_ndarray.cc:546: Operator _zeros is not implemented for GPU.

Stack trace returned 5 entries:
[bt] (0) 0   libmxnet.so                         0x0000000107a73358 _ZN4dmlc15LogMessageFatalD2Ev + 40
[bt] (1) 1   libmxnet.so                         0x000000010822f447 _Z20ImperativeInvokeImplRKN5mxnet7ContextEON4nnvm9NodeAttrsEPNSt3__16vectorINS_7NDArrayENS6_9allocatorIS8_EEEESC_PNS7_IbNS9_IbEEEESF_ + 2039
[bt] (2) 2   libmxnet.so                         0x00000001082304f7 MXImperativeInvoke + 439
[bt] (3) 3   libmxnet.so                         0x0000000108230ace MXImperativeInvokeEx + 46
[bt] (4) 4   _ctypes.so                          0x0000000106fd67d7 ffi_call_unix64 + 79

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/mxnet-0.11.1-py2.7.egg/mxnet/ndarray/sparse.py", line 123, in __repr__
    shape_info, self.context)
  File "/usr/local/lib/python2.7/site-packages/mxnet-0.11.1-py2.7.egg/mxnet/ndarray/ndarray.py", line 1147, in context
    return Context(Context.devtype2str[dev_typeid.value], dev_id.value)
KeyError: 0
>>>
```

Reply:
will be fixed in #7676